pkg/controllers/workspace_controller.go

```diff
@@ -149,10 +175,87 @@ func (c *WorkspaceReconciler) deleteWorkspace(ctx context.Context, wObj *kaitov1
 		klog.ErrorS(err, "failed to update workspace status", "workspace", klog.KObj(wObj))
 		return reconcile.Result{}, err
 	}
-
+	if err := c.deleteControllerRevisions(ctx, wObj); err != nil {
+		klog.ErrorS(err, "failed to delete ControllerRevisions", "workspace", klog.KObj(wObj))
+		return reconcile.Result{}, err
+	}
 	return c.garbageCollectWorkspace(ctx, wObj)
 }
 
+func (c *WorkspaceReconciler) deleteControllerRevisions(ctx context.Context, wObj *kaitov1alpha1.Workspace) error {
+	revisionList := &appsv1.ControllerRevisionList{}
+	if err := c.Client.List(ctx, revisionList, client.InNamespace(wObj.Namespace), client.MatchingLabels{
+		WorkspaceNameLabel: wObj.Name,
+	}); err != nil {
+		return fmt.Errorf("failed to list ControllerRevisions: %w", err)
+	}
+	for _, revision := range revisionList.Items {
+		if err := c.Client.Delete(ctx, &revision); err != nil {
+			return fmt.Errorf("failed to delete ControllerRevision %s: %w", revision.Name, err)
+		}
+	}
+	c.revisionCache = nil
+	return nil
+}
+
+func (c *WorkspaceReconciler) updateControllerRevision(ctx context.Context, wObj *kaitov1alpha1.Workspace) error {
+	needNewRevision := false
+	var currentHash string
+	if len(c.revisionCache) == 0 {
+		needNewRevision = true
+	} else {
+		currentHash = computeHash(wObj)
+		lastCachedHash := c.revisionCache[len(c.revisionCache)-1]
+		if currentHash != lastCachedHash {
+			needNewRevision = true
+		}
+	}
+	if !needNewRevision {
+		return nil
+	}
+	if currentHash == "" {
+		currentHash = computeHash(wObj)
+	}
+	data := map[string]string{"hash": currentHash}
+	jsonData, err := json.Marshal(data)
+	if err != nil {
+		return fmt.Errorf("failed to marshal revision data: %w", err)
+	}
+	newRevision := &appsv1.ControllerRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-%s", wObj.Name, currentHash[:8]),
+			Namespace: wObj.Namespace,
+			Labels: map[string]string{
+				WorkspaceRevisionLabel: "true",
+				WorkspaceNameLabel:     wObj.Name,
+			},
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(wObj, kaitov1alpha1.GroupVersion.WithKind("Workspace")),
+			},
+		},
+		Revision: int64(len(c.revisionCache) + 1),
+		Data:     runtime.RawExtension{Raw: jsonData},
+	}
+	if err := c.Client.Create(ctx, newRevision); err != nil {
+		return fmt.Errorf("failed to create new ControllerRevision: %w", err)
+	}
+	c.revisionCache = append(c.revisionCache, currentHash)
+	const maxRevisionHistoryLimit = 10
+	if len(c.revisionCache) > maxRevisionHistoryLimit {
+		c.revisionCache = c.revisionCache[1:]
+	}
```

Comment:
Can be replaced with something that adds the "controller-revision-hash" to the annotations and then updates the object.